clients/orderbook/src/worker.rs

```diff
@@ -0,0 +1,1289 @@
+						return Err(Error::StateHashMisMatch)
+					}
+					data.append(&mut chunk);
+				}
+			}
+			self.load_state_from_data(&data, summary)?;
+		}
+		Ok(())
+	}
+
+	// Checks if we have all stids to drive the state and then drive it.
+	pub async fn check_stid_gap_fill(&mut self) -> Result<(), Error> {
+		let mut last_snapshot = self.last_snapshot.read().state_change_id.saturating_add(1);
+
+		while let Some(action) = self.known_messages.remove(&last_snapshot) {
+			if let Err(err) = self.handle_action(&action) {
+				match err {
+					Error::Keystore(_) =>
+						error!(target:"orderbook","📒 BLS session key not found: {:?}",err),
+					_ => {
+						error!(target:"orderbook","📒 Error processing action: {:?}",err);
+						// The node found an error during processing of the action. This means we
+						// need to snapshot and drop everything else
+						self.snapshot(action.stid)?;
+						// We forget about everything else from cache.
+						self.known_messages.clear();
+						break
+					},
+				}
+			}
+			metric_set!(self, ob_state_id, last_snapshot);
+			last_snapshot = last_snapshot.saturating_add(1);
+		}
+		// We need to sub 1 since that last processed is one stid less than the not available
+		// when while loop is broken
+		self.last_snapshot.write().state_change_id = last_snapshot.saturating_sub(1);
+		Ok(())
+	}
+
+	pub fn want_stid(&mut self, from: &u64, to: &u64, peer: Option<PeerId>) {
+		if let Some(peer) = peer {
+			let mut messages = vec![];
+			for stid in *from..=*to {
+				// We dont allow gossip messsages to be greater than 10MB
+				if messages.encoded_size() >= 10 * 1024 * 1024 {
+					// If we reach size limit, we send data in chunks of 10MB.
+					let message = GossipMessage::Stid(messages);
+					self.gossip_engine.send_message(vec![peer], message.encode());
+					metric_inc!(self, ob_messages_sent);
+					metric_add!(self, ob_data_sent, message.encoded_size() as u64);
+					messages = vec![] // Reset the buffer
+				}
+				if let Some(msg) = self.known_messages.get(&stid) {
+					messages.push(msg.clone());
+				}
+			}
+			// Send the final chunk if any
+			if !messages.is_empty() {
+				let message = GossipMessage::Stid(messages);
+				self.gossip_engine.send_message(vec![peer], message.encode());
+				metric_inc!(self, ob_messages_sent);
+				metric_add!(self, ob_data_sent, message.encoded_size() as u64);
+			}
+		}
+	}
+
+	pub async fn got_stids_via_gossip(&mut self, messages: &Vec<ObMessage>) -> Result<(), Error> {
+		for message in messages {
+			// TODO: DO signature checks here and handle reputation
+			self.known_messages.entry(message.stid).or_insert(message.clone());
+		}
+		self.check_stid_gap_fill().await
+	}
+
+	// Expects the set bits in the bitmap to be missing chunks
+	pub async fn want(&mut self, snapshot_id: &u64, bitmap: &Vec<u128>, remote: Option<PeerId>) {
+		// Only respond if we are a fullnode
+		// TODO: Should we respond if we are also syncing???
```

Comment:
Still pending

Reply:
It's not pending, and these are something to test in testnet and then tweak.